### PR TITLE
[Fix] Refactor development reproducibility

### PR DESCRIPTION
**Pull request by franchuterivera**
FIx reproducibility in AutoPyTorch.

This is limited to CPU runs, as GPU based still will have non-deterministic algorithms enabled to take advantage of speed.

### CODE REVIEW

File: autoPyTorch/pipeline/components/training/data_loader/base_data_loader.py [hunk 1]
```diff
@@ -31,9 +31,11 @@ class BaseDataLoaderComponent(autoPyTorchTrainingComponent):
 
     """
 
-    def __init__(self, batch_size: int = 64) -> None:
+    def __init__(self, batch_size: int = 64,
+                 random_state: Optional[np.random.RandomState] = None) -> None:
         super().__init__()
         self.batch_size = batch_size
+        self.random_state = random_state
```
Comment (ravinkohli) [inline autoPyTorch/pipeline/components/training/data_loader/base_data_loader.py:38]: should we not send this random_state to the super class? So it is handled similar to how it is handled in the base component

File: autoPyTorch/pipeline/components/training/trainer/base_trainer.py [hunk 2]
```diff
@@ -161,9 +163,15 @@ def repr_last_epoch(self) -> str:
 
 class BaseTrainerComponent(autoPyTorchTrainingComponent):
 
-    def __init__(self, random_state: Optional[Union[np.random.RandomState, int]] = None) -> None:
+    def __init__(self, random_state: Optional[np.random.RandomState] = None) -> None:
         super().__init__()
-        self.random_state = random_state
+        if random_state is None:
```
Comment (ravinkohli) [inline autoPyTorch/pipeline/components/training/trainer/base_trainer.py:168]: same here? Why are we not passing the random state to the super class?

**Review by ravinkohli — COMMENTED** [review]
hey, Thank you for this PR. I have just a minor question, after it I resolved for me this PR can be merged

File: autoPyTorch/pipeline/components/training/trainer/base_trainer.py (hunk 2 above)
Comment (franchuterivera) [reply autoPyTorch/pipeline/components/training/trainer/base_trainer.py:167]: Sorry, I missed this as the base trainer relies on the torch manual seed. 

Anyhow, it makes sense to pass it also to the constructor and I have pushed a fix for this. Thanks for the catch.

**Review by ravinkohli — APPROVED** [review]

**Comment by ravinkohli** [comment]
I would also like to link this PR to its corresponding issue. #137

**Comment by ArlindKadra** [comment]
This is something that we need to handle quite fast. I will go through the PR and review it, then do we merge it to refactor development and we rebase the cocktails branch there?

@ravinkohli @franchuterivera

**Comment by ravinkohli** [comment]
> This is something that we need to handle quite fast. I will go through the PR and review it, then do we merge it to refactor development and we rebase the cocktails branch there?
> 
> @ravinkohli @franchuterivera

Yes, I think so. We should rebase the cocktails branch soon as we might start diverging quite a lot between the two branches.

**Review by ArlindKadra — COMMENTED** [review]
A few minor comments but overall really nice PR. Quite necessary regarding the reproducibility of the results :)

File: autoPyTorch/api/base_task.py [hunk 3]
```diff
@@ -1235,7 +1235,8 @@ def __del__(self) -> None:
         # When a multiprocessing work is done, the
         # objects are deleted. We don't want to delete run areas
         # until the estimator is deleted
-        self._backend.context.delete_directories(force=False)
+        if hasattr(self, '_backend'):
+            self._backend.context.delete_directories(force=False)
```
[the remark this answers is not in the document]
Comment (franchuterivera) [reply autoPyTorch/api/base_task.py:1239]: Exactly, not related to this PR -- for reproducibility -- but for the reason, you are saying.

File: autoPyTorch/pipeline/components/training/data_loader/base_data_loader.py [hunk 4]
```diff
@@ -31,9 +31,11 @@ class BaseDataLoaderComponent(autoPyTorchTrainingComponent):
 
     """
 
-    def __init__(self, batch_size: int = 64) -> None:
-        super().__init__()
+    def __init__(self, batch_size: int = 64,
+                 random_state: Optional[np.random.RandomState] = None) -> None:
+        super().__init__(random_state=random_state)
         self.batch_size = batch_size
+        self.random_state = random_state
```
[the remark this answers is not in the document]
Comment (franchuterivera) [reply autoPyTorch/pipeline/components/training/data_loader/base_data_loader.py]: Yes, thanks for the catch!

**Pull request merged**